Refactoring

ctuguinay opened the pull request:
### Refactoring

This PR addresses the following issues: #91, #88, #83, #43.

**The main folders are now:**
- utils
- lines
- regions2d

**As opposed to before:**
- utils
- convert
- formats
- mask
- plot

**Refactoring notes:**
- In both regions2d and lines, plotting is still kept as a class function: `r2d.plot() `and `lines.plot()`.
- In both regions2d and lines, masking is kept outside of class: `regions2d_mask(r2d,....)` and `lines_mask(lines,.....)`.
- In both regions2d and lines, parsing is kept outside of class: `parse_regions_file(input_file)` and `parse_lines_file(input_file)`.
- We now always initialize regions2d and lines with appropriate .evr and .evl functions at initialization ALWAYS. This is enforced with an error if not initialized properly (incorrect .evr or .evl file).

**Other Changes Include:**
- Added line masking based on Valentina's example notebook with options for what to do with NA values and what interpolation method to use.

All tests currently pass.

@leewujung @valentina-s Feel free to review this when you are free.

leewujung commented on echoregions/lines/lines.py:
```diff
@@ -98,48 +76,54 @@ def swap_val(val: Union[int, float]) -> Union[int, float]:
 
-        regions = self.data if inplace else self.data.copy()
+        regions = self._data if inplace else self._data.copy()
         regions.loc[:] = regions.apply(replace_depth, axis=1)
```
I wonder if we can replace the whole `replace_depth`/`swap_val` thing by a lambda function, since what it does is to swap out -10000.99 with `self.nan_depth_value`.

leewujung replied:
Can also define `ECHOVIEW_NAN_DEPTH_VALUE = -10000.99` near the top of the script and use this constant in place of the value, so that it is easier to keep track of numbers like this flying around.

See example here: https://github.com/OSOceanAcoustics/echopype/blob/9b401ca8250e60c7264b274e66145d0f7842def6/echopype/echodata/combine.py#L20-L26

leewujung commented on echoregions/lines/lines.py:
```diff
@@ -33,45 +37,19 @@ def __getitem__(self, idx: int) -> Union[Dict, List]:
-        """String path to the EVL file"""
-        return self._parser.input_file
-
     @property
     def nan_depth_value(self) -> Union[int, float]:
         return self._nan_depth_value
 
+    @property
+    def data(self) -> DataFrame:
+        return self._data
+
```
I don't see a lot of value in keeping `nan_depth_value` as a property -- what do you think about removing it? Also curious why you decided to have `.data` as a way to access `._data` -- is it for potential future changes?

leewujung commented on echoregions/lines/lines.py:
```diff
@@ -98,48 +76,54 @@ def swap_val(val: Union[int, float]) -> Union[int, float]:
-        if self.data is None:
-            self.parse_file(**kwargs)
-        self._parser.to_csv(self.data, save_path=save_path, **kwargs)
+        if not isinstance(self._data, DataFrame):
+            raise TypeError(
+                f"Invalid ds Type: {type(self._data)}. Must be of type DataFrame."
+            )
```
since `._data` is a private attribute, it seems having the type checking here is redundant and somewhat strange, since the type is set by the programmer

leewujung commented on echoregions/lines/lines.py:
```diff
@@ -1,29 +1,33 @@
-        self._masker = None
+class Lines:
+    def __init__(self, input_file: str, nan_depth_value: float = None):
+        self.depth = (
+            None  # Single array that can be used to obtain min_depth and max_depth
+        )
+        self._min_depth = (
+            None  # Set to replace -9999.99 depth values which are EVR min range
+        )
+        self._max_depth = (
+            None  # Set to replace 9999.99 depth values which are EVR max range
+        )
```
`self.depth`, `self._min_depth`, `self._max_depth` are not used anywhere in the Line class, so I am a little unsure why they are here. Am I missing anything?

leewujung commented on echoregions/lines/lines.py:
```diff
@@ -179,12 +163,14 @@ def plot(
                 f"start and end times are of type {type(start_time)} and {type(end_time)}. \
                             They must be of of type Pandas Timestamp."
             )
```
better to check each of these separately so the error message can be more specific.

leewujung commented on echoregions/__init__.py:
```diff
@@ -1,12 +1,21 @@
+from .utils.io import check_file, from_JSON, validate_path
+from .utils.time import parse_simrad_fname_time, parse_time
 
 __all__ = [
     "read_evl",
     "read_evr",
     "parse_simrad_fname_time",
     "parse_time",
+    "from_JSON",
+    "check_file",
+    "validate_path",
     "Lines",
     "Regions2D",
+    "lines_mask",
+    "regions2d_mask",
 ]
```
I think the ones imported in this root level `__init__.py` are the one that users will use directly, such as doing 
```python
test_line = read_evl(SOME_EVL)
test_region = read_evr(SOME_EVR)
```
So I'd suggest to keep only the ones below:
```suggestion
__all__ = [
    "read_evl",
    "read_evr",
    "lines_mask",
    "regions2d_mask",
]
```
We probably should discuss again later (but soon) if we are happy with the way `lines_mask` and `regions2d_mask` are.

leewujung replied:
I think the ones imported in this root level `__init__.py` are the one that users will use directly, such as doing 
```python
test_line = read_evl(SOME_EVL)
test_region = read_evr(SOME_EVR)
```
So I'd suggest to keep only the ones below:
```suggestion
__all__ = [
    "read_evl",
    "read_evr",
    "lines_mask",
    "regions2d_mask",
]
```
We probably should discuss again later (but soon) if we are happy with the way `lines_mask` and `regions2d_mask` are.

leewujung commented on echoregions/lines/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from .lines_mask import lines_mask
+from .lines_parser import parse_line_file
+
+__all__ = ["Lines", "lines_mask", "parse_line_file"]
```
could remove `parse_line_file` since that is not used outside of this module.

leewujung commented on echoregions/regions2d/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from .regions2d_mask import regions2d_mask
+from .regions2d_parser import parse_regions_file
+
+__all__ = ["Regions2D", "regions2d_mask", "parse_regions_file"]
```
could remove `parse_regions_file` since it is not used outside of this module.

leewujung commented on echoregions/regions2d/regions2d.py:
```diff
@@ -1,59 +1,52 @@
-        self._parser = Regions2DParser(input_file)
-        self._plotter = None
-        self._masker = None
+        self.depth = (
+            None  # Single array that can be used to obtain min_depth and max_depth
+        )
+        self._min_depth = (
+            None  # Set to replace -9999.99 depth values which are EVR min range
+        )
+        self._max_depth = (
+            None  # Set to replace 9999.99 depth values which are EVR max range
+        )
```
I think we can remove `self.depth` to not have a "hidden" thing for setting the max and min depth. Similar to the case in Lines, I also wonder if we can simply keep min_depth and max_depth as attributes and not properties to make things simpler, since it is unlikely these values will be set by some complicated method.

leewujung commented on echoregions/regions2d/regions2d.py:
```diff
@@ -84,50 +77,28 @@ def min_depth(self, val: Union[int, float]) -> None:
-        if self.data is None:
-            self.parse_file(**kwargs)
-        self._parser.to_csv(self.data, save_path=save_path, **kwargs)
+        if not isinstance(self.data, DataFrame):
+            raise TypeError(
+                f"Invalid ds Type: {type(self.data)}. Must be of type DataFrame."
+            )
```
Similar to the Lines case, I don't think we need to check the type of `self.data` here since it is set internally.

leewujung commented on echoregions/regions2d/regions2d.py:
```diff
@@ -330,41 +275,30 @@ def convert_points(
-                raise ValueError(
-                    "Input file has not been parsed; call `parse_file` to parse."
-                )
+        def _swap_depth_edge(self, y: Union[int, float]) -> Union[int, float]:
```
I am really confused by this and `convert_points` -- I don't see it used anywhere in the repo. Not sure if I am missing anything though. The use of 9999.99 and -9999.99 suggests that there's some redundancy with `replace_nan_depth`. Do we need to keep the `convert_points` function?

leewujung reviewed (commented):
Hey @ctuguinay : I am just half way into reviewing this but wanted to note down things I've noticed. Many of these were from the old code and not related to the refactoring, so I'd suggest to only change those that are very small and put the more involved ones as separate issues that will be addressed later, so that we can merge this PR first, especially since you have other open PR that depends on this one.

- I made some comments about `select_region` in [a comment for your other PR](https://github.com/OSOceanAcoustics/echoregions/pull/97#issuecomment-1606547559) -- I think that comment can be addressed there so it does not clogged up this refactoring PR.
- For `Regions2d.replace_nan_depth`: looking at its use `regions_mask`, I think there is some bug there -- when `inplace=True`, `self.data` should be updated and nothing is returned. When `inplace=False`, this method can return a new dataframe from `self.data`. Am I missing anything here? -- Feel free to make this a separate issue to be addressed later so it does not clog up this refactoring PR.
- The docstring of `select_region` does not make sense
- Should we remove `convert_points` altogether? (see [this comment](https://github.com/OSOceanAcoustics/echoregions/pull/95/files#r1241488893))
- The input argument `region` in `select_region` and a few other methods seems would be better renamed as `region_id` since it is effectively used to subset the dataframe with the column `region_id`

leewujung commented on echoregions/regions2d/regions2d_mask.py:
```diff
@@ -0,0 +1,136 @@
+from .regions2d import Regions2D
+
+
+def regions2d_mask(
+    da_Sv: DataArray,
+    regions2d: Regions2D,
+    region_ids: List,
+    mask_var: str = None,
+    mask_labels=None,
+) -> DataArray:
```
We want this to be a method `Regions2d.mask`, so the `regions2d` input argument would just be the `self`. Note the `replace_nan_depth` has a bug in doing `inplace=True` that needs to get fixed so that the nan_depth is replaced in `self.data` (I have this in my previous comment).

leewujung commented on echoregions/lines/lines_mask.py:
```diff
@@ -0,0 +1,101 @@
+from .lines import Lines
+
+
+def lines_mask(
+    da_Sv: DataArray, lines: Lines, method: str = "nearest", limit_area: str = None
+):
```
Similar to the regions2d case, we want this to be a method `Lines.mask` and not as a separate function.

leewujung commented on echoregions/regions2d/regions2d_mask.py:
```diff
@@ -0,0 +1,136 @@
+
+    # Set up mask labels.
+    if mask_labels:
+        if mask_labels == "from_ids":
+            # Create mask.
+            r = regionmask.Regions(outlines=regions_np, numbers=region_ids)
+            M = r.mask(
+                da_Sv,
+                lon_name="unix_time",
+                lat_name="depth",
+                wrap_lon=False,
+            )
+
+        elif isinstance(mask_labels, list):
+            # Create mask.
+            r = regionmask.Regions(outlines=regions_np)
+            M = r.mask(
+                da_Sv,
+                lon_name="unix_time",
+                lat_name="depth",
+                wrap_lon=False,
+            )
+            # Convert default labels to mask_labels.
+            S = xr.where(~M.isnull(), 0, M)
+            S = M
+            for idx, label in enumerate(mask_labels):
+                S = xr.where(M == idx, label, S)
+            M = S
+        else:
+            raise ValueError("mask_labels must be None, 'from_ids', or a list.")
+    else:
+        # Create mask.
+        r = regionmask.Regions(outlines=regions_np)
+        M = r.mask(
+            da_Sv,
+            lon_name="unix_time",
+            lat_name="depth",
+            wrap_lon=False,
+        )
```
This entire block can be simplified since the `M = r.mask(...)` section is identical. The code will be much cleaner if it starts with a section for creating `r`, going into `M = r.mask(...)`, and then some conversion from labels to mask_labels if `mask_labels` is a list.

leewujung commented on echoregions/lines/lines_mask.py:
```diff
@@ -0,0 +1,101 @@
+    filtered_bottom = filter_bottom(lines_df, start_time, end_time)
+
+    if len(filtered_bottom) > 0:
+        # create joint index
+        joint_index = list(
+            set(list(pd.DataFrame(sonar_index)[0]) + list(filtered_bottom.index))
+        )
+
+        # interpolate on the sonar coordinates
+        # nearest interpolation has a problem when points are far from each other
+        bottom_interpolated = filtered_bottom.reindex(joint_index).loc[
+            sonar_index
+        ]  # .interpolate('nearest').loc[sonar_index]
```
I think one can do reindex directly to `sonar_index` without creating the `joint_index` first?

leewujung commented on echoregions/lines/lines_mask.py:
```diff
@@ -0,0 +1,101 @@
+                .interpolate(method=method, limit_area=limit_area)
+                .loc[sonar_index]
+            ).fillna(max_depth)
+        except:
+            print("")
+            raise ValueError(
+                "Interpolation arguments are invalid. Visit the docs at \
+                 https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.interpolate.html \
+                 for more information on what can be placed in said arguments."
+            )
```
remove `print("")`?

Probably a good idea to catch the invalid `method` and `limit_area` explicitly before the try-catch block, and use this block to catch other unexpected things. There's sort of a mismatch here since the function only allows `method` and `limit_area` but `interpolate` allows many more arguments, so the error string should also be revised. Probably better to just directly raise the error from `interpolate`.

leewujung commented on echoregions/utils/io.py:
```diff
@@ -0,0 +1,78 @@
+    return str(save_path)
+
+
+def check_file(file: str, format: Union[List[str], str]) -> None:
```
Suggest to change the function to be more specific
```suggestion
def check_file_type(file: str, format: Union[List[str], str]) -> None:
```

leewujung replied:
Also I think the content of this function should be cleaned up (the logic and also to check filename extension properly with `pathlib` or properly extracting the .XXX part instead of `endwith`). Could you put this as a separate issue?

leewujung reviewed (changes requested):
Hey @ctuguinay : Thanks for the refactoring work! I am done with this round of review. All comments are either in my overall comment last night or inline with the code. 

One last thing is the location of `read_evr` and `read_evl`. Could you move them out of `utils` and put them into `core.py` at root level? (`echoregions/echoregions/core.py`) This way we can easily find them, and __init__ is easy. We may decide to move them later to another module, but for now let's keep it simple.

ctuguinay commented:
@leewujung Sorry for how long this comment is. I mostly just wanted to keep all these replies to your comments in one place, for easier access later on.

**1**
> I wonder if we can replace the whole `replace_depth`/`swap_val` thing by a lambda function, since what it does is to swap out -10000.99 with `self.nan_depth_value`.

> Can also define `ECHOVIEW_NAN_DEPTH_VALUE = -10000.99` near the top of the script and use this constant in place of the value, so that it is easier to keep track of numbers like this flying around.
> 
> See example here: https://github.com/OSOceanAcoustics/echopype/blob/9b401ca8250e60c7264b274e66145d0f7842def6/echopype/echodata/combine.py#L20-L26

I addressed both of these in the latest push.

**2**
> I don't see a lot of value in keeping `nan_depth_value` as a property -- what do you think about removing it? Also curious why you decided to have `.data` as a way to access `._data` -- is it for potential future changes?

True, no reason keeping `nan_depth_value` as a property; I've removed it as a property (and so I had to remove its setter too) in the latest push. Also `.data` feels in sync with how xarray's data is accessed.

> since `._data` is a private attribute, it seems having the type checking here is redundant and somewhat strange, since the type is set by the programmer

True, I think I initially wrote this prior to making it a private attribute. I removed it in latest push.

**3**
> `self.depth`, `self._min_depth`, `self._max_depth` are not used anywhere in the Line class, so I am a little unsure why they are here. Am I missing anything?

Yup, removed them in latest push. Initially moved them here since they were inherited by a Geometry class that I deleted.

**4**
> better to check each of these separately so the error message can be more specific.

Just separated them, can be found in latest push.

**5**
> I think the ones imported in this root level `__init__.py` are the one that users will use directly, such as doing
> 
> ```python
> test_line = read_evl(SOME_EVL)
> test_region = read_evr(SOME_EVR)
> ```
> 
> So I'd suggest to keep only the ones below:
> .....
>
> We probably should discuss again later (but soon) if we are happy with the way `lines_mask` and `regions2d_mask` are.

Since I am placing both masking functions as class methods, I don't think they need to be here. So the only functions needed are read_evl and read_evr. I made this change and it can be found in the latest push.

**6**
> could remove `parse_line_file` since that is not used outside of this module.

> could remove `parse_regions_file` since it is not used outside of this module.

I addressed both of these in the latest push. For the reason listed above, I removed both mask functions too.

**7**
> I think we can remove `self.depth` to not have a "hidden" thing for setting the max and min depth. Similar to the case in Lines, I also wonder if we can simply keep `min_depth` and `max_depth` as attributes and not properties to make things simpler, since it is unlikely these values will be set by some complicated method.

Just removed `self.depth` and removed `min_depth` and `max_depth` as a property. Also I do agree with you that they would be more appropriate as an attribute. I will still keep the setter methods because of problems that may arise in initialization. These changes can be found in latest push.

**8**
> Similar to the Lines case, I don't think we need to check the type of `self.data` here since it is set internally.

Just removed this. Can be found in latest push.

**9**
> I am really confused by this and `convert_points` -- I don't see it used anywhere in the repo. Not sure if I am missing anything though. The use of 9999.99 and -9999.99 suggests that there's some redundancy with `replace_nan_depth`. Do we need to keep the `convert_points` function?

I will put this in a new issue and will reference it in a bit. I don't see it used anywhere, but it was created for a reason. It is related to #53, and as such may be used there.

**10**
> * For `Regions2d.replace_nan_depth`: looking at its use `regions_mask`, I think there is some bug there -- when `inplace=True`, `self.data` should be updated and nothing is returned. When `inplace=False`, this method can return a new dataframe from `self.data`. Am I missing anything here? -- Feel free to make this a separate issue to be addressed later so it does not clog up this refactoring PR.

This is what pandas copy is saying:

DataFrame.copy(deep=True)
Make a copy of this object’s indices and data.

When deep=True (default), a new object will be created with a copy of the calling object’s data and indices. Modifications to the data or indices of the copy will not be reflected in the original object (see notes below).

I'll place this in a separate issue to investigate further and mention this PR when I’ve created it. But yes, I think you're right in the sense that nothing should be returned if true, but in this case, if true it will return something instead of changing it inplace.

**11**
> * The input argument `region` in `select_region` and a few other methods seems would be better renamed as `region_id` since it is effectively used to subset the dataframe with the column `region_id`

I address this in #97, so I'll leave it in there as it's more relevant for that PR.

**12**
> * The input argument `files` in `select_sonar_file` seems would be better renamed as `sonar_filenames` to avoid confusion

Just changed this. Can be found in latest push.

**13**
> We want this to be a method `Regions2d.mask`, so the `regions2d` input argument would just be the `self`. Note the `replace_nan_depth` has a bug in doing `inplace=True` that needs to get fixed so that the nan_depth is replaced in `self.data` (I have this in my previous comment).
> 
> So effectively we will not have a separate `regions2d_mask.py` and will access this method via the `Regions2d` objects.

> Similar to the regions2d case, we want this to be a method `Lines.mask` and not as a separate function.

Just made them class functions. Can check the latest push.

**14**
> This entire block can be simplified since the `M = r.mask(...)` section is identical. The code will be much cleaner if it starts with a section for creating `r`, going into `M = r.mask(...)`, and then some conversion from labels to mask_labels if `mask_labels` is a list.

Due to how large this PR is, I'll place this in a separate issue and mention this PR when I’ve created it.

**15**
> I think one can do reindex directly to `sonar_index` without creating the `joint_index` first?

Not sure why, but reindexing directly to `sonar_index` led to key not found errors, I did a preliminary search, but didn’t find anything that would cause this discrepancy other than an index reset after .loc[...]. Do you want me to investigate this further, or just leave it?

**16**
> remove `print("")`?
> 
> Probably a good idea to catch the invalid `method` and `limit_area` explicitly before the try-catch block, and use this block to catch other unexpected things. There's sort of a mismatch here since the function only allows `method` and `limit_area` but `interpolate` allows many more arguments, so the error string should also be revised. Probably better to just directly raise the error from `interpolate`.

Just changed this code. Can be found in latest push.

**17**
> For cloud compatibility (see #69) we would want this changed to something that is more comprehensive. The same validate function in echopype probably can just be grabbed directly without much change to work here. Could you please make a note in the #69 comment so that we do not forget about this?

Just made a note!

**18**
> Also I think the content of this function should be cleaned up (the logic and also to check filename extension properly with `pathlib` or properly extracting the .XXX part instead of `endwith`; `pathlib` does not work for cloud path last time I checked so maybe some simple `re` use or similar). Could you put this as a separate issue?

I will place this in a separate issue and mention this PR when I’ve created it.

**19**
> One last thing is the location of `read_evr` and `read_evl`. Could you move them out of `utils` and put them into `core.py` at root level? (`echoregions/echoregions/core.py`) This way we can easily find them, and **init** is easy. We may decide to move them later to another module, but for now let's keep it simple.

Just moved them. Can be found in latest push.

leewujung commented:
> > I think one can do reindex directly to `sonar_index` without creating the `joint_index` first?
> 
> Not sure why, but reindexing directly to `sonar_index` led to key not found errors, I did a preliminary search, but didn’t find anything that would cause this discrepancy other than an index reset after .loc[...]. Do you want me to investigate this further, or just leave it?

Could you put in a TODO comment in the code to say what you have here, and we can potentially investigate later? It runs through fine so no reason for this to be a blocker. Thanks!

leewujung commented on echoregions/lines/lines.py:
```diff
@@ -174,3 +151,114 @@ def plot(
+        Subsets a bottom dataset to the range of an Sv dataset
+        Create a mask of same shape as data found in Sonar object; bottom: False, otherwise: True
+
+        Arguments:
+            sonar - Sonar object containing DataArray of shape (ping_time, depth).
+            method - String containing interpolation method.
+            limit_area - String for determining filling restriction for NA values.
+
+        Returns:
+        bottom_mask - xarray with dimensions: (ping_time, depth) with bottom: False, otherwise: True
```
Let's also change this to numpy docstring format to be consistent.

leewujung reviewed (approved):
@ctuguinay : Thanks for making the changes and adding issues for the various pieces.

I only have a minor comment about docstring in one of the mask functions, but other than that, feel free to merge this PR!
